io.fabric8.kubernetes.client.informers.impl.cache.Reflector.ReflectorWatcher#reconnecting() {  return true;

### Describe the bug

io.fabric8.kubernetes.client.dsl.internal.AbstractWatchManager#cannotReconnect（）， always returns false，this will cause the client to keep trying to establish a connection asynchronously



### Fabric8 Kubernetes Client version

6.4.1

### Steps to reproduce

k8s ApiServer rate limit

### Expected behavior

Cancel asynchronous task



### Runtime

Kubernetes (vanilla)

### Kubernetes API Server version

other (please specify in additional context)

### Environment

Linux

### Fabric8 Kubernetes Client Logs

```shell
2023-07-11 18:00:48.727 DEBUG -- [pool-2-thread-3] .i.AbstractWatchManager : Scheduling reconnect task
2023-07-11 18:00:48.727 DEBUG -- [pool-2-thread-3] ckoffIntervalCalculator : Current reconnect backoff is 32000 milliseconds (T336)
2023-07-11 18:00:49.678 DEBUG -- [138.36:6443/...] i.f.k.c.i.i.c.Reflector : Event received MODIFIED Service resourceVersion v1536952902 for v1/services
2023-07-11 18:00:49.678 DEBUG -- [138.36:6443/...] i.f.k.c.i.i.c.Reflector : Event received MODIFIED Service resourceVersion v1536952902 for v1/services
2023-07-11 18:00:50.082 DEBUG -- [ool-2-thread-35] .i.AbstractWatchManager : Watching https://11.50.138.174:6443/api/v1/services?resourceVersion=152868113&allowWatchBookmarks=true&watch=true...
2023-07-11 18:00:50.083 DEBUG -- [ool-2-thread-35] i.f.k.client.Config     : Trying to configure client from Kubernetes config...
2023-07-11 18:00:50.083 DEBUG -- [ool-2-thread-35] i.f.k.client.Config     : Did not find Kubernetes config at: [/root/.kube/config]. Ignoring.
2023-07-11 18:00:50.083 DEBUG -- [ool-2-thread-35] i.f.k.client.Config     : Trying to configure client from service account...
2023-07-11 18:00:50.083 DEBUG -- [ool-2-thread-35] i.f.k.client.Config     : Found service account host and port: 10.233.0.1:443
2023-07-11 18:00:50.083 DEBUG -- [ool-2-thread-35] i.f.k.client.Config     : Found service account ca cert at: [/var/run/secrets/kubernetes.io/serviceaccount/ca.crt}].
2023-07-11 18:00:50.083 DEBUG -- [ool-2-thread-35] i.f.k.client.Config     : Found service account token at: [/var/run/secrets/kubernetes.io/serviceaccount/token].
2023-07-11 18:00:50.083 DEBUG -- [ool-2-thread-35] i.f.k.client.Config     : Trying to configure client namespace from Kubernetes service account namespace path...
2023-07-11 18:00:50.083 DEBUG -- [ool-2-thread-35] i.f.k.client.Config     : Found service account namespace at: [/var/run/secrets/kubernetes.io/serviceaccount/namespace].
2023-07-11 18:00:50.083 DEBUG -- [ool-2-thread-35] atcherWebSocketListener : WebSocket error received
java.util.concurrent.CompletionException: java.io.InterruptedIOException: executor rejected
        at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:332)
        at java.base/java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:1177)
        at java.base/java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:2341)
        at io.fabric8.kubernetes.client.http.StandardHttpClient.buildWebSocket(StandardHttpClient.java:128)
        at io.fabric8.kubernetes.client.http.StandardWebSocketBuilder.buildAsync(StandardWebSocketBuilder.java:43)
        at io.fabric8.kubernetes.client.dsl.internal.WatchConnectionManager.start(WatchConnectionManager.java:113)
        at io.fabric8.kubernetes.client.dsl.internal.AbstractWatchManager.startWatch(AbstractWatchManager.java:224)
        at io.fabric8.kubernetes.client.dsl.internal.AbstractWatchManager.reconnect(AbstractWatchManager.java:151)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
        at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
        at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:735)
        at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:732)
        at java.base/java.security.AccessController.doPrivileged(AccessController.java:400)
        at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1.run(Executors.java:732)
        at java.base/java.lang.Thread.run(Thread.java:1570)
Caused by: java.io.InterruptedIOException: executor rejected
        at okhttp3.RealCall$AsyncCall.executeOn(RealCall.java:157)
```


### Additional context

_No response_

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -8,6 +8,7 @@
 * Fix #5423: OkHttpClientImpl supports setting request method for empty payload requests
 
 #### Improvements
+* Fix #5327: added proactive shutdown of informers on client close
 * Fix #5432: [java-generator] Add the possibility to always emit `additionalProperties` on generated POJOs
 * Fix #5368: added support for additional ListOptions fields
 * Fix #5377: added a createOr and unlock function to provide a straight-forward replacement for createOrReplace.

--- a/httpclient-okhttp/src/main/java/io/fabric8/kubernetes/client/okhttp/OkHttpClientImpl.java
+++ b/httpclient-okhttp/src/main/java/io/fabric8/kubernetes/client/okhttp/OkHttpClientImpl.java
@@ -55,9 +55,7 @@ import org.slf4j.LoggerFactory;
 import java.io.BufferedInputStream;
 import java.io.IOException;
 import java.io.InterruptedIOException;
-import java.io.PrintWriter;
 import java.io.Reader;
-import java.io.StringWriter;
 import java.lang.reflect.Method;
 import java.net.MalformedURLException;
 import java.nio.ByteBuffer;
@@ -250,15 +248,6 @@ public class OkHttpClientImpl extends StandardHttpClient<OkHttpClientImpl, OkHtt
 
   @Override
   public void close() {
-    if (LOG.isDebugEnabled()) {
-      StringWriter writer = new StringWriter();
-      PrintWriter printWriter = new PrintWriter(writer);
-      new Exception().printStackTrace(printWriter);
-      printWriter.close();
-      String stack = writer.toString();
-      stack = stack.substring(stack.indexOf("\n"));
-      LOG.debug("Shutting down dispatcher {} at the following call stack: {}", this.httpClient.dispatcher(), stack);
-    }
     ConnectionPool connectionPool = httpClient.connectionPool();
 
     Dispatcher dispatcher = httpClient.dispatcher();

--- a/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/BaseOperation.java
+++ b/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/BaseOperation.java
@@ -51,6 +51,7 @@ import io.fabric8.kubernetes.client.dsl.base.PatchContext;
 import io.fabric8.kubernetes.client.dsl.base.PatchType;
 import io.fabric8.kubernetes.client.extension.ExtensibleResource;
 import io.fabric8.kubernetes.client.http.HttpRequest;
+import io.fabric8.kubernetes.client.impl.BaseClient;
 import io.fabric8.kubernetes.client.informers.ResourceEventHandler;
 import io.fabric8.kubernetes.client.informers.SharedIndexInformer;
 import io.fabric8.kubernetes.client.informers.impl.DefaultSharedIndexInformer;
@@ -1037,6 +1038,7 @@ public class BaseOperation<T extends HasMetadata, L extends KubernetesResourceLi
     if (indexers != null) {
       informer.addIndexers(indexers);
     }
+    this.context.getClient().adapt(BaseClient.class).getClosed().whenComplete((closed, ignored) -> informer.stop());
     return informer;
   }
 

--- a/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/impl/BaseClient.java
+++ b/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/impl/BaseClient.java
@@ -41,17 +41,22 @@ import io.fabric8.kubernetes.client.http.HttpClient;
 import io.fabric8.kubernetes.client.utils.ApiVersionUtil;
 import io.fabric8.kubernetes.client.utils.KubernetesSerialization;
 import io.fabric8.kubernetes.client.utils.Utils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 import java.net.HttpURLConnection;
 import java.net.MalformedURLException;
 import java.net.URL;
 import java.util.List;
+import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.Executor;
 import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Executors;
 import java.util.function.Predicate;
 
 public abstract class BaseClient implements Client {
+
+  public static final Logger logger = LoggerFactory.getLogger(BaseClient.class);
 
   /**
    * An {@link ExecutorSupplier} that provides an unlimited thread pool {@link Executor} per client.
@@ -84,10 +89,12 @@ public abstract class BaseClient implements Client {
   private ExecutorSupplier executorSupplier;
   private Executor executor;
   protected KubernetesSerialization kubernetesSerialization;
+  private CompletableFuture<Void> closed;
 
   private OperationContext operationContext;
 
   BaseClient(BaseClient baseClient) {
+    this.closed = baseClient.closed;
     this.config = baseClient.config;
     this.httpClient = baseClient.httpClient;
     this.adapters = baseClient.adapters;
@@ -104,6 +111,7 @@ public abstract class BaseClient implements Client {
 
   BaseClient(final HttpClient httpClient, Config config, ExecutorSupplier executorSupplier,
       KubernetesSerialization kubernetesSerialization) {
+    this.closed = new CompletableFuture<>();
     this.config = config;
     this.httpClient = httpClient;
     this.handlers = new Handlers();
@@ -136,11 +144,20 @@ public abstract class BaseClient implements Client {
 
   @Override
   public synchronized void close() {
+    if (closed.complete(null) && logger.isDebugEnabled()) {
+      logger.debug(
+          "The client and associated httpclient {} have been closed, the usage of this or any client using the httpclient will not work after this",
+          httpClient.getClass().getName());
+    }
     httpClient.close();
     if (this.executorSupplier != null) {
       this.executorSupplier.onClose(executor);
       this.executorSupplier = null;
     }
+  }
+
+  public CompletableFuture<Void> getClosed() {
+    return closed;
   }
 
   @Override

--- a/kubernetes-tests/src/test/java/io/fabric8/kubernetes/client/mock/DefaultSharedIndexInformerTest.java
+++ b/kubernetes-tests/src/test/java/io/fabric8/kubernetes/client/mock/DefaultSharedIndexInformerTest.java
@@ -1204,6 +1204,23 @@ class DefaultSharedIndexInformerTest {
     assertEquals(0, foundExistingAnimal.getCount());
   }
 
+  @Test
+  void testClientStopClosesInformer() throws InterruptedException {
+    // Given
+    setupMockServerExpectations(Animal.class, "ns1", this::getList,
+        r -> new WatchEvent(getAnimal("red-panda", "Carnivora", r), "ADDED"), null, null);
+
+    // When
+    SharedIndexInformer<GenericKubernetesResource> animalSharedIndexInformer = client
+        .genericKubernetesResources(animalContext)
+        .inNamespace("ns1")
+        .runnableInformer(60 * WATCH_EVENT_EMIT_TIME);
+
+    client.close();
+
+    assertTrue(animalSharedIndexInformer.stopped().toCompletableFuture().isDone());
+  }
+
   private KubernetesResource getAnimal(String name, String order, String resourceVersion) {
     AnimalSpec animalSpec = new AnimalSpec();
     animalSpec.setOrder(order);

